### ISSUE TITLE
Stripping QueryRoot in SailQuery

### ISSUE THREAD

**Issue by frensjan** [eclipse-rdf4j/rdf4j#4859]
### Problem description

For query optimisation it is useful to add a `QueryRoot` node above a tuple expression. This is what e.g. happens in `SailSourceConnection#evaluateInternal`.

However there is I think an issue in the interaction with the explain feature. If a `Connection` adds a query root and the original root is replaced by an optimisation, the explanation has a reference to the original root, not to the replaced root and because of it the wrong plan is shown.

At the same time, in `SailQuery#explain` the query to explain is modified to _not_ include a `QueryRoot`.

### Preferred solution

I would prefer that the `QueryRoot` is not removed in `SailQuery#explain` and that this is instead left up to the `SailConnection`.

### Are you interested in contributing a solution yourself?

Yes

**Comment by hmottestad** [comment]
Maybe the explain feature should actually add a QueryRoot? 

I think that maybe @JervenBolleman stripped the QueryRoot in the explain code so that one of the tests didn't fail, but from the looks of it the test org.eclipse.rdf4j.sail.memory.QueryPlanRetrievalTest#testWildcard might actually be wrong.

**Comment by frensjan** [comment]
> Maybe the explain feature should actually add a QueryRoot?

Maybe? I'd say it'll probably suffice to tackle that in the `Connection` though. `SailConnection#explain` returns a `o.e.rdf4j.query.explanation.Explanation`, so there shouldn't be any issues with the root `TupleExpr` being altered without that being visible on the outside.

**Comment by JervenBolleman** [comment]
My intention was that there should always be a queryroot for evaluation ;) but as all queries have a root, I didn't think that it would be useful for an explanation. 

Replacing the root is going to be strange anyway as you don't have a reference to the new root outside the optimizer?

**Comment by frensjan** [comment]
I have the impression that the reason for having a `QueryRoot` is to create a sort out _out_ parameter. 

This is used in e.g. [SailSourceConnection#explain](https://github.com/eclipse-rdf4j/rdf4j/blob/a1973a3f7b89ea173f4cbee5d8de9547c14ca418/core/sail/base/src/main/java/org/eclipse/rdf4j/sail/base/SailSourceConnection.java#L282). It relies on the `TupleExpr` to be 'explained' being modified by `runQueryForExplain` and subsequently `evaluate` and `evaluateInternal`.

Now, [SPARQLParser#parseQuery](https://github.com/eclipse-rdf4j/rdf4j/blob/a1973a3f7b89ea173f4cbee5d8de9547c14ca418/core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/SPARQLParser.java#L209) already adds a `QueryRoot`, although I can imagine that on other places you can't rely on this so the occasional _adding_ of a `QueryRoot` if it isn't there is a good thing. _Especially_ in `evaluateInternal` as otherwise it may not be possible to 'optimise' the (original) root.

But _removing_ a `QueryRoot` in `SailQuery#explain` is inconvenient of the (original) root is being replaced by an optimiser in `DefaultEvaluationStrategy#optimize`.

A practical example is a query such as this:
```sparql
select ?entity
where {
  ?entity :name "Frens" .
  ?entity :age ?age .
}
order by desc(?age)
limit 3
```
which maps to the following algebra:
```
 (slice _ 3
   (project (?entity)
     (order ((desc ?age))
       (bgp
         (triple ?entity :name "Frens")
         (triple ?entity :age ?age)
       ))))))
```
can be completely optimised in our system with a single but different (!) operator, pushed down into the underlying storage.

_However_, explain doesn't register this if there is no `QueryRoot` used as a reference from `SailQuery#explain` to `SailConnection#explain`.

**Comment by frensjan** [comment]
TL;DR 😄 

Could we drop
```java
if (tupleExpr instanceof QueryRoot) {
	tupleExpr = ((QueryRoot) tupleExpr).getArg();
}
```

from `SailQuery`?

**Comment by hmottestad** [comment]
> TL;DR 😄 
> 
> 
> 
> Could we drop
> 
> ```java
> 
> if (tupleExpr instanceof QueryRoot) {
> 
> 	tupleExpr = ((QueryRoot) tupleExpr).getArg();
> 
> }
> 
> ```
> 
> 
> 
> from `SailQuery`?


Can you link to a line in the file in GitHub?

**Comment by frensjan** [comment]
It's in the `if` block on [line 47 in SailQuery.java](https://github.com/eclipse-rdf4j/rdf4j/blob/a1973a3f7b89ea173f4cbee5d8de9547c14ca418/core/repository/sail/src/main/java/org/eclipse/rdf4j/repository/sail/SailQuery.java#L47).

**Comment by hmottestad** [comment]
I'm good with that. I might even go as far as to check for a QueryRoot and add one if it's missing, since the QueryRoot is what allows the query explanation to work if the top query node is replaced during optimisation.

**Comment by frensjan** [comment]
I've created [a PR](https://github.com/eclipse-rdf4j/rdf4j/pull/4862).
